### ISSUE TITLE
Improve compatibility with major libraries

### ISSUE THREAD

**Issue by proofconstruction** [sparkfish/augraphy#108]
# Background
Augraphy's library API is currently the [odd man out](https://www.merriam-webster.com/dictionary/odd%20man%20out) in the image degradation space, and since we're one of the only projects doing *document* image degradation, this is even more apparent.

## ML Libraries
[Fastai pipelines](https://fastcore.fast.ai/transform#Pipeline), [Tensorflow datasets](https://www.tensorflow.org/tfx/transform/get_started), and [PyTorch datasets/dataloaders](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html) all expect callables (functions or things that act like them) as transforms. These accept a single image as argument and return the transformed image. Albumentations transforms and imgaug augmenters are already callable and behave this way, but Augraphy augmentations do not.

## Image Degradation Libraries
[`Albumentations.Compose`](https://github.com/albumentations-team/albumentations/blob/master/albumentations/core/composition.py#L134) and [`imgaug.augmenters.Sequential`](https://github.com/aleju/imgaug/blob/master/imgaug/augmenters/meta.py#L3006) objects are wrappers for lists of transforms, similar to our `AugmentationSequence`, and are also callables like ours, but these always return a transformed image instead of putting it into a data dictionary and passing that around like we do.

# The Problem
Just for reference, [here's an example of how Albumentations is used with PyTorch.](https://albumentations.ai/docs/examples/pytorch_classification/)

In the section **"Use Albumentations to define transformation functions for the train and validation datasets"**, the example defines a transformation as an Albumentations Compose, and uses it when creating a dataset object:

```python
val_transform = A.Compose(
    [
        A.SmallestMaxSize(max_size=160),
        A.CenterCrop(height=128, width=128),
        A.Normalize(mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)),
        ToTensorV2(),
    ]
)
val_dataset = CatsVsDogsDataset(images_filepaths=val_images_filepaths, transform=val_transform)
```

Ignoring that we wouldn't ever use the Cats and Dogs dataset with Augraphy, doing the same thing with our library would require defining an AugraphyPipeline, then writing a function which accepts an image argument, calls the pipeline on the image, and returns `pipeline.augment(image)['output']`, like this:

```python
ink = [Letterpress()]

paper = [PaperFactory(),
              BrightnessTexturize()]

post = [BadPhotoCopy()]

pipeline = AugraphyPipeline(ink,paper,post)

def val_transform(img):
    augmented = pipeline.augment(img)
    return augmented['output']

val_dataset = CatsVsDogsDataset(images_filepaths=val_images_filepaths, transform=val_transform)
```

This would be necessary **even if** someone only wanted to use one augmentation, because all augmentations interact with the data dictionary and thus require the presence of a full AugraphyPipeline.

# A Solution
## Proposal
To fit with the usage patterns and interface expectations of these other major libraries, I suggest we:
1. move responsibility for appending `AugmentationResult`s to the data dictionary up into the Augraphy pipeline itself,
2. convert all augmentations and `AugmentationSequence` to callables that accept an image and return an image,
3. add a `__call__` method to `AugraphyPipeline` that performs `self.augment()` and returns the output image from `data['output']`, so full pipelines could be used as transforms, allowing users to train models on certain kinds of document effects, as mentioned [here](https://github.com/sparkfish/augraphy/issues/33#issuecomment-916573343).

## Outcomes
This will have several benefits:
1. Augmentations are a little bit easier to write; they basically become fancy functions that have internal state defined in their constructors. Augmentations would no longer deal with retrieving the most recent image, nor storing the result of their computations back in the data dictionary; this would be handled in the pipeline.
2. Augmentations are also easier to debug, because we could use them directly *or* in a pipeline, instead of *only* in a pipeline, opening up some more possibilities for unit-testing, as well as just making it straightforward to shim inputs.
3. Callable `AugmentationSequence`s are immediately intuitive for anyone using `Albumentations.Compose` or `imgaug.augmenters.Sequential`.
4. Single augmentations, `AugmentationSequence`s, and entire `AugraphyPipelines` would be usable directly in Fastai, Tensorflow, and PyTorch applications.
5. Users who want to inspect the data dictionary can still access it by using an `AugraphyPipeline`, but anyone who just wants to consume the library can use augmentations or `AugmentationSequence`s directly.

I don't see any downsides beyond a rather small refactor and updating some documentation, a lot of which I already did as part of the [Composability](https://github.com/sparkfish/augraphy/issues/22) branch.

**Comment by jboarman** [comment]
Looks excellent @proofconstruction.

Regardless of the loss of a potentially useful feature of the current pipeline, I think the proposal as offered is necessary.

At its core, the proposal is basically to “convert all augmentations and AugmentationSequence to callables that accept an image and return an image”.

This approach means that we lose our current ability to track any state between augmentations and allow for any interactions between steps of a sequence.  That said, to my knowledge, this feature has remained only theoretical and not acted upon at this point.  So no immediate “harm” should result from his change.  And, if we find a compelling reason to re-add this functionality, I’m certain we can assess the cost of a clever solution for whatever the perceived benefits at the time would be, all while retaining interoperability with other common frameworks.

**Comment by proofconstruction** [comment]
[This commit](https://github.com/sparkfish/augraphy/commit/d0c6107229c56afdef04e572377f96ba58f6584e) contains the changes.

# Result
Users can now apply augmentations

```python
img = cv2.imread('rauchen.png')
markup = Markup()
marked = markup(img)
```
![marked](https://user-images.githubusercontent.com/74747193/136653811-6c9db6f4-73e3-43e8-be19-89f9808e3d79.png)

or augmentation sequences
```python
markup_then_fold = AugmentationSequence([Markup(), Folding()])
marked_and_folded = markup_then_fold(img)
```
![marked_and_folded](https://user-images.githubusercontent.com/74747193/136653823-40be02ca-b0c7-4747-9970-f3e11dccdd79.png)

directly, like functions.

# Summary of Changes
## Pipeline
Created an `apply_phase` method which takes the ink, paper, or post phases and applies every augmentation in them, if any, to the most recent image in that layer. This method also handles recording execution time and storing `AugmentationResult`s in the pipeline `data` dictionary. Layers only make sense in the context of a pipeline, so storing augmentations in layers now only happens here; augmentations don't know about the layer they're executed in.

## AugmentationSequence
The `__call__` method now applies every augmentation in the sequence, then returns the result.

## OneOf
The `__call__` method selects one augmentation in the list of augmentations passed to it, executes it on the input image, then returns the result. This has a 100% chance of occurring. To *possibly* but not *necessarily* execute an augmentation inside a `OneOf`, best-practice is to give `OneOf` a probability not equal to 1. For example, the following code will always execute either `Brightness` or `Gamma`:

```python
OneOf([Brightness(), Gamma()])
```
while this code has a 50% chance of executing **one of** `Brightness` or `Gamma`, and a 50% chance of doing nothing at all:

```python
OneOf([Brightness(), Gamma()], p=0.5)
```
## PaperFactory
If called, `PaperFactory` now returns the paper, after tiling the texture. It does not return the texture, nor does it record the texture in the data dictionary. This is a slight feature regression, and we can discuss how to fix it, if desired.

## Augmentations
Augmentations are all now callable on images and return the transformed image. Augmentations no longer know about the layer they're called in, when they are called as part of a pipeline. Augmentations  don't know about the `data` dictionary either, as this is also handled in the pipeline, or not at all if they are used individually or as part of an `AugmentationSequence`. Augmentations **do** now take a `layer` argument in their `__call__` method, which defaults to `None` (so users don't need to worry about this when using them as functions as part of a sequence). When used in a pipeline, the `apply_phase` method passes in the correct string for the `layer` parameter. This allows augmentations to still have layer-specific behavior. For example, the `Brightness` augmentation can be called in any layer but has specific behavior when called in the Ink layer: it performs additional image-space transformations to and from grayscale.

## Default Pipeline
The default Augraphy pipeline in `augraphy.default.pipeline` was updated to reflect the changes to augmentation probability and the removal of the layer parameter.

## Other
1. Because all augmentations, `OneOf`, and `AugmentationSequence` may now be called as functions, their probabilities default to 1, to guarantee their execution in that context.
2. I made some minor edits to the `__call__` methods to match our agreed-upon style in #94. Mostly this looks like changing `img` to `image`, but I also changed the name of a few functions in `augraphy.augmentations.lib` from `thisStyle` to `this_style`. For example, `addNoise` became `add_noise`. We can make other edits like this as part of future PRs. We couldn't agree on the short `img_blurred` vs long `image_blurred` style for variables and parameters, so I think a good compromise is this: use the long form in the `__call__` and `__init__` methods, because that's what people will read first to figure out what each augmentation actually does. Everywhere else, like in auxiliary functions, the short form is fine.

**Comment by proofconstruction** [comment]
Documentation has been slightly updated to reflect these changes, version 5 has been cut and released on PyPI.
